src/main/java/fi/aalto/cs/apluscourses/intellij/model/IntelliJModelFactory.java

```diff
@@ -22,11 +28,43 @@ public IntelliJModelFactory(@NotNull Project project) {
   public Course createCourse(@NotNull String name,
                              @NotNull List<Module> modules,
                              @NotNull Map<String, String> requiredPlugins) {
-    return new IntelliJCourse(name, modules, requiredPlugins, project);
+    IntelliJCourse course = new IntelliJCourse(name, modules, requiredPlugins, project);
+    // Add a module change listener with the created course instance to the project
+    project.getMessageBus().connect().subscribe(ProjectTopics.MODULES, new ModuleListener() {
+      @Override
+      public void moduleRemoved(@NotNull Project project,
+                                @NotNull com.intellij.openapi.module.Module projectModule) {
+        Module module = course.getModuleOpt(projectModule.getName());
+        if (module != null) {
+          course.onModuleRemove(module);
+        }
+      }
+    });
+    project.getMessageBus().connect().subscribe(VirtualFileManager.VFS_CHANGES,
+        new BulkFileListener() {
+          @Override
+          public void after(@NotNull List<? extends VFileEvent> events) {
+            for (VFileEvent event : events) {
+              if (event instanceof VFileDeleteEvent) {
+                String deletedFileName = event.getFile().getName();
+                Module module = course.getModuleOpt(deletedFileName);
+                if (module != null) {
+                  course.onModuleFilesDeletion(module);
+                }
+              }
+            }
+          }
+        }
+    );
+    return course;
   }
 
   @Override
   public Module createModule(@NotNull String name, @NotNull URL url) {
-    return new IntelliJModule(name, url, project);
+    Module module = new IntelliJModule(name, url, project);
+    // IntelliJ modules may already be present in the project or file system, so we update the state
+    // at module creation here.
+    module.updateState();
+    return module;
```

Comment:
I slightly disagree with semantics here... the name of the method does not exactly reflect the contents: _create_ vs. _update_. I'd propose a simpler renaming from `createModule()` into something like `updateOrOtherwiseCreate()`